blockchain/validate.go

```diff
@@ -710,11 +719,22 @@ func CheckBlockHeaderContext(header *wire.BlockHeader, prevNode HeaderCtx,
 			str = fmt.Sprintf(str, header.Timestamp, medianTime)
 			return ruleError(ErrTimeTooOld, str)
 		}
-	}
 
-	// The height of this block is one more than the referenced previous
-	// block.
-	blockHeight := prevNode.Height() + 1
+		// Testnet4 only: Check timestamp against prev for difficulty-adjustment
+		// blocks to prevent timewarp attacks.
+		if c.ChainParams().EnforceBIP94 {
+			// Check timestamp for the first block of each difficulty adjustment
+			// interval, except the genesis block.
+			if blockHeight%c.BlocksPerRetarget() == 0 {
+				prevBlockTimestamp := time.Unix(prevNode.Timestamp(), 0)
+				if header.Timestamp.Before(prevBlockTimestamp.Add(-maxTimeWarp)) {
+					str := "block's timestamp %v is too early on diff adjustment block %v"
+					str = fmt.Sprintf(str, header.Timestamp, prevBlockTimestamp)
+					return ruleError(ErrTimewarpAttack, str)
+				}
+			}
+		}
+	}
```

Comment:
nit: line length. Also, sounds like this whole block could be extracted into a standalone function. Something like `checkBip94()`.